subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PreCompiledScriptRunner.java

```diff
@@ -0,0 +1,175 @@
+        this.serviceRegistry = serviceRegistry;
+
+        this.scriptRunnerFactory = serviceRegistry.get(ScriptRunnerFactory.class);
+        this.scriptHandlerFactory = serviceRegistry.get(ScriptHandlerFactory.class);
+        this.pluginRequestApplicator = serviceRegistry.get(PluginRequestApplicator.class);
+
+        this.classLoaderScope = classLoaderScope.createChild("pre-compiled-script").lock();
+        this.pluginManager = pluginManager;
+    }
+
+    public void run(Class<?> pluginsBlockClass, Class<?> scriptClass) {
+        if (pluginsBlockClass != null) {
+            applyPlugins(pluginsBlockClass);
+        }
+
+        if (scriptClass != null) {
+            executeScript(scriptClass);
+        }
+    }
+
+    private void applyPlugins(Class<?> pluginsBlockClass) {
+        ScriptSource scriptSource = new PrecompiledScriptSource(pluginsBlockClass);
+        ScriptRunner<PluginsAwareScript, ?> runner = scriptRunnerFactory.create(compiledPluginsBlock(pluginsBlockClass), scriptSource, classLoaderScope.getExportClassLoader());
+        runner.run(target, serviceRegistry);
+
+        ScriptHandlerInternal scriptHandler = scriptHandlerFactory.create(scriptSource, classLoaderScope);
+        pluginRequestApplicator.applyPlugins(runner.getScript().getPluginRequests(), scriptHandler, pluginManager, classLoaderScope);
+    }
+
+    private CompiledScript<PluginsAwareScript, ?> compiledPluginsBlock(Class<?> pluginsBlockClass) {
+        return new CompiledGroovyPlugin<>(pluginsBlockClass, PluginsAwareScript.class);
+    }
+
+    private void executeScript(Class<?> scriptClass) {
+        ScriptSource scriptSource = new PrecompiledScriptSource(scriptClass);
+        scriptRunnerFactory.create(compiledScript(scriptClass), scriptSource, classLoaderScope.getExportClassLoader())
+            .run(target, serviceRegistry);
+    }
+
+    private CompiledScript<BasicScript, ?> compiledScript(Class<?> precompiledScriptClass) {
+        return new CompiledGroovyPlugin<>(precompiledScriptClass, BasicScript.class);
+    }
+
+    private class CompiledGroovyPlugin<T extends Script> implements CompiledScript<T, Object> {
+
+        private final Class<? extends T> compiledClass;
+
+        private CompiledGroovyPlugin(Class<?> scriptClass, Class<T> scriptBaseClass) {
+            this.compiledClass = scriptClass.asSubclass(scriptBaseClass);
+        }
+
+        @Override
+        public boolean getRunDoesSomething() {
+            return true;
+        }
+
+        @Override
+        public boolean getHasMethods() {
+            return false;
+        }
+
+        @Override
+        public Class<? extends T> loadClass() {
+            return compiledClass;
+        }
+
+        @Override
+        public Object getData() {
+            return null;
+        }
+
+        @Override
+        public void onReuse() {
+            classLoaderScope.onReuse();
+        }
+    }
+
+    private static class PrecompiledScriptSource implements ScriptSource {
```

Comment:
❓ is this new implementation of `ScriptSource` necessary?

Reply:
Turned out to be not necessary. Removed.